hera_cal/smooth_cal.py

```diff
@@ -171,247 +133,128 @@ def time_filter(gains, wgts, times, filter_scale = 120.0, nMirrors = 0):
+            calfits_list: list of string paths to calfits files containing calibration solutions and flags
+            flags_npz_list: list of string paths to npz files containing flags as a function of baseline, times
+                and frequency. Must have all baselines for all times. Flags on baselines are broadcast to both 
+                antennas involved, unless either antenna is completely flagged for all times and frequencies.
         '''
-        assert(isinstance(cal, (str, UVCal)))
-        self.input_cal = cal
-        self.gains, self.flags, self.quals, _, _, self.freqs, self.times, self.pols =  io.load_cal(cal, return_meta=True)
-        self.nFreq = len(self.freqs)
-        self.nInt =len(self.times)
-        self.tInt = np.median(np.diff(self.times))*24.0*60.0*60.0
-        self.has_cal = True
+        # load calibration files
+        self.cals = sorted(calfits_list)
+        self.gains, self.cal_flags, self.cal_freqs, self.cal_times = odict(), odict(), odict(), odict()
+        for cal in self.cals:
+            (self.gains[cal], self.cal_flags[cal], _, _, _, self.cal_freqs[cal], 
+                 self.cal_times[cal], _) = io.load_cal(cal, return_meta=True)
+            
+        # load flags files
+        self.npzs = sorted(flags_npz_list)
+        if len(self.npzs) > 0:
+            self.npz_flags, self.npz_freqs, self.npz_times = odict(), odict(), odict()
+            for cal, npz in zip(self.cals, self.npzs):
+                self.npz_flags[cal] = synthesize_ant_flags(io.load_npz_flags(npz))
+                npz_dict = np.load(npz)
+                self.npz_freqs[cal] = np.unique(npz_dict['freq_array'])
+                self.npz_times[cal] = np.unique(npz_dict['time_array'])
+                    
+        # set up time grid
+        all_file_times = np.array(self.cal_times.values()).ravel()
+        self.dt = np.median(np.diff(all_file_times))
+        self.time_grid = np.arange(all_file_times[0], all_file_times[-1]+self.dt, self.dt)
+        self.time_indices = {cal: np.searchsorted(self.time_grid, times) for cal, times in self.cal_times.items()}
+                
+        # build multi-file grids for each antenna's gains and flags
+        self.freqs = self.cal_freqs[self.cals[0]]
+        self.ants = sorted(list(set([k for gain in self.gains.values() for k in gain.keys()])))
+        self.gain_grids = {ant: np.ones((len(self.time_grid), len(self.freqs)), 
+                                        dtype=np.complex) for ant in self.ants}
+        self.flag_grids = {ant: np.ones((len(self.time_grid), len(self.freqs)), 
+                                        dtype=bool) for ant in self.ants}
+        for ant in self.ants:
+            for cal in self.cals:
+                if self.gains[cal].has_key(ant):
+                    self.gain_grids[ant][self.time_indices[cal],:] = self.gains[cal][ant]
+                    self.flag_grids[ant][self.time_indices[cal],:] = self.cal_flags[cal][ant]
```

Comment:
The way this is written precludes the possibility of a non-one-to-one mapping from gains to time_indices.  Would be better to use something like, e.g., aipy.utils.add2array to handle both cases.  otherwise, as written, there's a non-deterministic mapping of gains into grids in the multi-to-one mapping case.

Reply:
Let's just enforce one-to-one mapping for now.